### PR TITLE
DateOffset add to ndarray

### LINKED ISSUE #47014: ENH: implement DateOffset + datetime64 array
#### Is your feature request related to a problem?

DateOffset + a numpy array of type datetime64 does not work, see examples at the end of this post

#### Describe the solution you'd like

can `__add__` or `__radd__` of DateOffset be changed to implement addition to datetime64 array?

#### API breaking implications

I am not the right person to comment, but seems unlikely to have an impact on existing API if it can be done at all

#### Describe alternatives you've considered

Sorry, not thought of alternatives so far

#### Code example

```python
>>> import pandas
>>> import numpy
>>> dates = pandas.Series([pandas.Timestamp(2021, 1, 1) + pandas.Timedelta(days=d) for d in range(1000)])
>>> offset = pandas.offsets.MonthEnd()

>>> # adding a pandas Timestamp series to an offset obviously works
>>> dates + offset
0     2021-01-31
1     2021-01-31
         ...    
998   2023-09-30
999   2023-09-30
Length: 1000, dtype: datetime64[ns]

>>> # adding the underlying numpy datetime64 values does not work 
>>> offset + dates.values
---------------------------------------------------------------------------
UFuncTypeError                            Traceback (most recent call last)
----> 1 offset + dates.values

UFuncTypeError: ufunc 'add' cannot use operands with types dtype('O') and dtype('<M8[ns]')

>>> # slightly different error if we swap the order of the two variables
>>> dates.values + offset
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
----> 1 dates.values + offset

TypeError: Concatenation operation is not implemented for NumPy arrays, use np.concatenate() instead. Please do not rely on this error; it may not be given on all Python implementations.

>>> # the desired result can be achieved by using DateOffset._apply_array() and casting to M8[ns]
>>> offset._apply_array(dates.values).astype('M8[ns]')
array(['2021-01-31T00:00:00.000000000', '2021-01-31T00:00:00.000000000',
       '2021-01-31T00:00:00.000000000', '2021-01-31T00:00:00.000000000',
       '2021-01-31T00:00:00.000000000', '2021-01-31T00:00:00.000000000',
       '2021-01-31T00:00:00.000000000', '2021-01-31T00:00:00.000000000',
       etc])
```

### PR DESCRIPTION
**Pull request by JonAnCla**
This is WIP to implement changes needed to allow DateOffset subclasses e.g. MonthEnd to be directly added to numpy.ndarray, e.g. `numpy.array(['2022-01-01'], dtype='datetime64') + pandas.offsets.MonthEnd(1)`

I need some guidance on:

- cython mechanism to check that passed ndarray is datetime64 dtype i.e.cython equivalent of pandas.api.types.is_datetime64_any_dtype
- how to approach adding unit tests? two ideas I have
  - somehow duplicate existing offset tests but run them on ndarray instead of Series?  looks on the surface to be more difficult to implement
  - apply a set of offsets to a predefined Series and ndarray equivalent, and then check results match

Many thanks!

- [x] closes #47014
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/_libs/tslibs/offsets.pyx [hunk 1]
```diff
@@ -456,8 +456,12 @@ cdef class BaseOffset:
         return type(self)(n=1, normalize=self.normalize, **self.kwds)
 
     def __add__(self, other):
-        if util.is_array(other) and other.dtype == object:
-            return np.array([self + x for x in other])
+        if util.is_array(other):
+            from pandas.api.types import is_datetime64_any_dtype
+            if other.dtype == object:
+                return np.array([self + x for x in other])
+            elif is_datetime64_any_dtype(other):
```
Comment (mroeschke) [inline pandas/_libs/tslibs/offsets.pyx:463]:
```suggestion
            if other.dtype == object:
                return np.array([self + x for x in other])
            elif other.dtype.kind == "M":
```

File: pandas/_libs/tslibs/offsets.pyx [hunk 2]
```diff
@@ -456,8 +456,12 @@ cdef class BaseOffset:
...
+                return self._apply_array(other).astype(other.dtype)
```
Comment (mroeschke) [inline pandas/_libs/tslibs/offsets.pyx:464]: Do we need the astype here?

**Review by mroeschke — CHANGES_REQUESTED** [review]
This would also need tests and a whatsnew note in `v3.0.0.rst`

File: pandas/_libs/tslibs/offsets.pyx (hunk 2 above)
Comment (hottwaj) [reply pandas/_libs/tslibs/offsets.pyx:463]: I found that without this DateOffset._apply_array did not respect the dtype was passed in
once added in, I could use e.g. datetime64[D] and get a datetime64[D] back

I'll do some further testing when working on unit tests

Comment (mroeschke) [reply pandas/_libs/tslibs/offsets.pyx:463]: FWIW `datetime64[D]` is not a supported unit in pandas

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

**Pull request closed without merging**